### DISCUSSION TITLE
Using PyTorch, CuPy and PyQtGraph for visualizing Images on the GPU

### DISCUSSION THREAD

**Discussion by adossantos21** [pyqtgraph/pyqtgraph#2652] [Q&A]
Hi, I'm trying to display/visualize an image that is of type cupy.ndarray. The project I'm working on loads an image using PIL Image.open and converts it to a PyTorch torch.tensor on the CUDA Device. That torch.tensor is then copied to the CPU where it is converted to a numpy.ndarray. This conversion is necessary for our team to visualize the image (since a numpy array is used as input to pyqtgraph.ImageItem.setImage). For this project, we **need** to visualize the image. Currently, I'm looking for a way to visualize the image without copying the data to the host (CPU). 
I was hoping that CuPy might help with this task (as we want to perform operations on the CUDA device). I've edited the global configuration options in PyQtGraph and set useCupy and useOpenGL to True with pyqtgraph.setConfigOptions(). However, while performing tests using a CuPy array vs a NumPy array, results have shown that using a CuPy array is not beneficial for speeding up visualization of the image. I cannot disclose the full code my team is working on nor explain why we convert from a PIL format to torch.tensor to numpy.ndarray, but I have displayed two simple test programs that mimic what we've tried to do. I was hoping someone could explain why the NumPy array visualization runs faster than the CuPy array visualization? I suspect that when we plot the data, it is copied back to the device but am not sure. Does anyone know if there is a way to visualize an image without copying the data to the CPU? We want to avoid the copy to the CPU so we can speed up the process of visualizing the image (by visualizing it on the GPU). Apologies if this question is a dumb one, I'm just out of ideas at this point. I have PyQt5 and PySide2 installed via pip. I am running on an x86_64 architecture. Please let me know if there is any other information about my system you would like me to include.

Here is the code for the GPU CuPy version:
`
from datetime import datetime
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QWidget, QGraphicsWidget
import sys
import cupy as cp
import numpy as np
import torch
from pyqtgraph import GraphicsItem, GraphicsWidget, PlotItem
import pyqtgraph as pg
from PIL import Image
import torchvision.transforms as transforms

# GPU Version
def main():

        # Initialize QApplication and configure CuPy
        print("\n CuPy: ", pg.getConfigOption('useCupy'),"\n")
        pg.setConfigOption('useCupy', 'True')
        print("\n CuPy: ", pg.getConfigOption('useCupy'),"\n")
        print("\n OpenGL: ", pg.getConfigOption('useOpenGL'), "\n")
        pg.setConfigOption('useOpenGL', 'True')
        print("\n OpenGL: ", pg.getConfigOption('useOpenGL'), "\n")
        application = QApplication(sys.argv)

        # Start Program
        startTime = datetime.now()    # to time the program process
        image = Image.open("sheep.png")

        # Define a transform to convert PIL
        # image to a Torch tensor
        transform = transforms.Compose([
        transforms.PILToTensor()
        ])

        # Convert the PIL image to Torch tensor
        img_tensor = transform(image)
        cuda0 = torch.device('cuda:0')
        img_tensor = img_tensor.to(cuda0)
        # Convert image tensor to cupy array
        img_gpu = cp.asarray(img_tensor)
       startTime2 = datetime.now()    # to observe the operations after cupy array was created

        # Set pyqtgraph configuration properties and try to display an image
        img_pg = pg.ImageItem()
        img_pg.setImage(img_gpu.T)
        plot_pg = pg.plot()
        plot_pg.addItem(img_pg)
        QtCore.QTimer.singleShot(5*1000, QtCore.QCoreApplication.quit) # this closes the widget window after 5 seconds so the runtime can be computed
        print(datetime.now() - startTime)
        print(datetime.now() - startTime2)

        sys.exit(application.exec_())

if __name__ == '__main__':
        main()
`
Here is the code for the CPU NumPy version:
`
from datetime import datetime
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QWidget, QGraphicsWidget
import sys
import cupy as cp
import numpy as np
import torch
from pyqtgraph import GraphicsItem, GraphicsWidget, PlotItem
import pyqtgraph as pg
from PIL import Image
import torchvision.transforms as transforms

# CPU Version
def main():

        # Initialize QApplication and check numpy or cupy config option
        print("\n CuPy: ", pg.getConfigOption('useCupy'),"\n")
        print("\n OpenGL: ", pg.getConfigOption('useOpenGL'), "\n")
        application = QApplication(sys.argv)

        # Start Program
        startTime = datetime.now()    #to time the program process
        image = Image.open("sheep.png")

        # Define a transform to convert PIL
        # image to a Torch tensor
        transform = transforms.Compose([
        transforms.PILToTensor()
        ])

        # Convert the PIL image to Torch tensor
        img_tensor = transform(image)
        cuda0 = torch.device('cuda:0')
        img_tensor = img_tensor.to(cuda0)
        # Convert image tensor to NumPy array
        img_tensor = img_tensor.cpu() # I understand the redundancy of copying to cuda then copying back to cpu, but my team's real program does this for a specific reason
        img_cpu = np.asarray(img_tensor)
       startTime2 = datetime.now()    # to observe the operations after numpy array was created

        # Set pyqtgraph configuration properties and try to display an image
        img_pg = pg.ImageItem()
        img_pg.setImage(img_cpu.T)
        plot_pg = pg.plot()
        plot_pg.addItem(img_pg)
        QtCore.QTimer.singleShot(5*1000, QtCore.QCoreApplication.quit) # this closes the widget window after 5 seconds so the runtime can be computed
        print(datetime.now() - startTime)
        print(datetime.now() - startTime2)

        sys.exit(application.exec_())

if __name__ == '__main__':
        main()
`
The image I load, "sheep.png" can be replaced with any png image. It is an arbitrary image I grabbed from the internet.
As previously stated, I'm not sure why the CPU NumPy Version runs faster than the GPU CuPy Version. My results from datetime are as follows:
CPU NumPy Version RunTime: 
    First datetime timer:       0:00:01.864312
    Second datetime timer:  0:00:00.057221

GPU CuPy Version RunTime:
    First datetime timer:       0:00:02.028029
    Second datetime timer:  0:00:00.155967

Here is the terminal output for running each program. The CuPy version is titled gpuTest.py and the NumPy version is titled cpuTest.py:
`
(pyqtEnv) [user1@rhel]$ python3 cpuTest.py

 CuPy:  False 

 OpenGL:  False 

QStandardPaths: error creating runtime directory '/run/user/578770' (Permission denied)
0:00:01.864312
0:00:00.057221
~/.local/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:263: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  "torch.distributed.reduce_op is deprecated, please use "
(pyqtEnv) [user1@rhel]$ python3 gpuTest.py

 CuPy:  False 

 CuPy:  True 

 OpenGL:  False 

 OpenGL:  True 

QStandardPaths: error creating runtime directory '/run/user/578770' (Permission denied)
0:00:02.028029
0:00:00.155967
~/.local/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:263: UserWarning: torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead
  "torch.distributed.reduce_op is deprecated, please use "
`
Thank you to anyone that reaches out and again, sorry if this is a dumb question!

**Comment by adossantos21** [comment]
Apologies, this is my first discussion post. I really didn't mean to mess up the format. Please let me know if it isn't decipherable and I'll try again.

**Reply by pijyoi** [reply]
Some points regarding use of ImageItem:
1) there's no need to set useOpenGL=True. I would recommend not setting it.
2) instantiate ImageItem with axisOrder='row-major' and pass in the image without transposing it. 
3) there's VideoSpeedTest.py in the examples, you can try it out with various configurations in an interactive manner to see how various parameters influence the performance

Question: why is it a "problem" that the cpu version runs faster than the gpu version?

**Reply by adossantos21** [reply]
Hi, thanks for the tips! We already have a script that utilizes the cpu version. It’s not a problem, it just wasn’t the expected outcome (given that CuPy was supposed to speed things up). The hope was that keeping the operations on the GPU and not copying to the cpu would speed up the process of displaying the image.

**Reply by j9ac9k** [reply]
How large are the images you're working with? We have observed some performance penalties when working with small images in CuPy, but the relative performance improves with image size.

**Reply by pijyoi** [reply]
There is a generic code path that converts the data to floating point and eventually creates a 4-channel ARGB32 QImage for display by Qt.

The cupy code path always goes through this generic code path. 

The cpu code path short circuits this code path for the commonly used image formats and can be very efficient.

**Reply by adossantos21** [reply]
@j9ac9k The images I'm working with are 640 x 480, but the test image I used for this script was 800 x 800.

**Reply by adossantos21** [reply]
@pijyoi Then it makes sense why the cpu numpy code path is more efficient. Can I ask where you found the information that told you the code paths the cpu and cupy go through?

**Reply by adossantos21** [reply]
@j9ac9k @pijyoi I was under the assumption that trying to keep operations on the gpu would speed my processes - for visualizing the image - up. Am I under the wrong assumption, or is there another CUDA enhanced library I might be able to explore to achieve my goals? I'm not looking for a handout, if you can point me in the right direction I can do some investigation and testing. Also just to be explicitly clear, when I say "goals", I mean utilizing the GPU for faster image visualization preferably without copying to host. I previously tried gpumat from OpenCV, but it doesn't take CuPy arrays as input or a torch.tensor. Thank you both for the quick responses.

**Reply by pijyoi** [reply]
The code is in ImageItem.py::render().

However, I was mistaken about cupy version always taking the generic code path. In fact both cupy and numpy versions will do short circuiting for commonly used image formats. 

An RGBA png image would qualify for short circuiting, and almost no processing is done on it by ImageItem.

**Reply by j9ac9k** [reply]
> @j9ac9k The images I'm working with are 640 x 480, but the test image I used for this script was 800 x 800.

We did a paper a while ago, didn't get accepted, but it has some figures on performance improvement for using cupy vs. numpy for images:

https://drive.google.com/file/d/11Zs-AKtJjlwm_0VxMavfzAC_RL07k2B7/view?usp=sharing

Here is the relevant figure: https://github.com/pyqtgraph/IEEEVIS2021-pyqtgraph/blob/main/figures/makeARGB_benchmark.pdf

There is a little variability depending if you have a LUT and what types you're data is, but cupy doesn't start getting faster for you until your images get to be larger than 10^5 - 10^6 pixels.  It's not surprising to me that cupy is slower for images 800x800 and smaller.


That said, you mentioned your data was _already_ in a cupy array, and I always theorized that the bottleneck, was getting the data to the GPU to begin with, but perhaps there is some optimization in pyqtgraph to be had (or perhaps my assumption was wrong).

> @pijyoi Then it makes sense why the cpu numpy code path is more efficient. Can I ask where you found the information that told you the code paths the cpu and cupy go through?

@pijyoi is more or less singlehandedly implemented all the different code paths for optimization, nobody knows the differences of the different code paths more :)


> @j9ac9k @pijyoi I was under the assumption that trying to keep operations on the gpu would speed my processes - for visualizing the image - up. Am I under the wrong assumption, or is there another CUDA enhanced library I might be able to explore to achieve my goals? I'm not looking for a handout, if you can point me in the right direction I can do some investigation and testing. Also just to be explicitly clear, when I say "goals", I mean utilizing the GPU for faster image visualization preferably without copying to host. I previously tried gpumat from OpenCV, but it doesn't take CuPy arrays as input or a torch.tensor. Thank you both for the quick responses.

AFAIK we haven't done much (any?) testing when pyqtgraph ingests cupy arrays to begin with.  It's quite possible we have some unnecessary conversions happening between numpy arrays and cupy arrays.  @outofculture is the individual that largely brought cupy support to the library, I don't typically work on systems with CUDA installed so I have only a little experience myself.

I suppose the first thing to do would be to create a minimal example that takes in already created cupy arrays and go from there.

If you're looking for hardware-accelerated render speed; VisPy may be another library to look at, although I'm not sure to what degree they support CUDA, but they use a lot of highly optimized opengl.

**Reply by pijyoi** [reply]
A google search for "cupy opengl interop" gives:
https://gist.github.com/keckj/e37d312128eac8c5fca790ce1e7fc437

That's what displaying cuda data without transferring back to cpu would entail.

**Reply by pijyoi** [reply]
So the right place to implement "cupy opengl interop" might perhaps be `RawImageWidget.RawImageGLWidget`. Instead of uploading a texture from cpu memory, I suppose you would use the cuda buffer and avoid the gpu memory -> cpu memory -> gpu memory round trip.

**Reply by pijyoi** [reply]
#3377 has an implementation of the cuda interops with OpenGL texture.
